runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -4758,10 +4763,34 @@ J9::Z::TreeEvaluator::multianewArrayEvaluator(TR::Node * node, TR::CodeGenerator
 
    bool use64BitClasses = comp->target().is64Bit() && !TR::Compiler->om.generateCompressedObjectHeaders();
 
-   // Init class field, then jump to end of ICF
    generateRXInstruction(cg, use64BitClasses ? TR::InstOpCode::STG : TR::InstOpCode::ST, node, classReg, generateS390MemoryReference(targetReg, TR::Compiler->om.offsetOfObjectVftField(), cg));
-   cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
-   iComment("Init class field and jump to end of ICF.");
+#if defined(TR_TARGET_64BIT)
+   if (isIndexableDataAddrPresent)
+      {
+      TR_ASSERT_FATAL_WITH_NODE(node,
+         ((TR::Compiler->om.compressObjectReferences()
+               && (fej9->getOffsetOfDiscontiguousDataAddrField() - fej9->getOffsetOfContiguousDataAddrField()) == 8)
+            || (!TR::Compiler->om.compressObjectReferences()
+               && fej9->getOffsetOfDiscontiguousDataAddrField() == fej9->getOffsetOfContiguousDataAddrField()),
+         "Offset of dataAddr field in discontiguous array is expected to be 8 bytes more than contiguous array if using compressed refs, "
+         "Or same if using full refs. "
+         "But was %d bytes for discontigous and %d bytes for contiguous array.\n",
+         fej9->getOffsetOfDiscontiguousDataAddrField(), fej9->getOffsetOfContiguousDataAddrField()));
+
+      // Load dataAddr slot offset difference since 0 size arrays are treated as discontiguous.
+      generateRILInstruction(cg,
+         TR::InstOpCode::LGFI,
```

Comment:
Minor improvement... If the offset difference is ever only going to be 0 or 8, we can use `LGHI` instead of `LGFI` and save 2-bytes in i-text.